h2o-core/src/main/java/hex/grid/filter/KeepOnlyFirstMatchFilterFunction.java

```diff
@@ -11,29 +11,36 @@
  *  Usage: pass the function to the constructor which will return `true` for the grid items of type <Map<String, Object>
  *  that we will want to consider equal.
  */
-public class KeepOnlyFirstMatchFilterFunction implements PermutationFilterFunction {
+public class KeepOnlyFirstMatchFilterFunction<MP extends Model.Parameters> implements PermutationFilterFunction<MP> {
 
-  // Function that should return `true` for grid items that we want to evaluate only one representative from.
-  public final Function<Map<String, Object>, Boolean> _baseMatchFunction;
+  // Function that should return `true` for permutations that we want to evaluate only one representative from.
+  public final Function<MP, Boolean> _baseMatchFunction;
   private int _maxNumberOfMatchesToApply = 1;
   
-  public KeepOnlyFirstMatchFilterFunction(Function<Map<String, Object>, Boolean> fun) {
+  public KeepOnlyFirstMatchFilterFunction(Function<MP, Boolean> fun) {
```

Comment:
This one is still confusing me when I look at how it's used in the tests.
So, it always matches (even when the passed function doesn't, which I the weirdest part..), except if it re-matches. Correct?
It's really hard to find a good name for such a filter.
The current name makes you think that it will match only the first match, but also that (as it's usually the case) it won't match when the function doesn't... but the difference is important here.
What about "MatchFirstOrTrueFilter" ?